coderd/database/migrations/000205_unique_linked_id.up.sql

```diff
@@ -0,0 +1,18 @@
+-- Remove the linked_id if two user_links share the same value.
+-- This will affect the user if they attempt to change their settings on
+-- the oauth/oidc provider. However, if two users exist with the same
+-- linked_value, there is no way to determine correctly which user should
+-- be updated. Since the linked_id is empty, this value will be linked
+-- by email.
+UPDATE ONLY user_links out
+SET
+	linked_id =
+		CASE WHEN (
+			  -- When the count of linked_id is greater than 1, set the linked_id to empty
+			  SELECT  COUNT(*)
+			  FROM    user_links inn
+			  WHERE   out.linked_id = inn.linked_id AND out.login_type = inn.login_type
```

Comment:
If we want to be really explicit about only matching this specific case, should we also check `out.user_id != inn.user_id`? This obviously has the chance of failing on index creation if this constraint isn't met.

What about checking oauth expiry? Only zero the one with an older expiry? Don't know if that's what we want, just putting ideas out there.

Reply:
The index is only on `login_type` and `linked_id`. The issue at present is that 2 users can share the same `linked_id`. If I add a `user_id` check, then the counts would be per user, and I could still have a constraint error.

> What about checking oauth expiry? Only zero the one with an older expiry? Don't know if that's what we want, just putting ideas out there.

Nah. We _should_ only have 1 user_link per user. Although maybe that is a better migration to apply first. I have that in an issue here: https://github.com/coder/coder/issues/12116

Fixing that is a more challenging problem though, because you'd have to decide which login is the correct one.